File: CuttingPolygonsC/src/polygon.cpp
#include "polygon.h"

namespace PolygonNamespace
{

Polygon::Polygon()
{
    _polygon.resize(3);
    for(int i=0; i<3;i++)
    {
        _polygon[i]=Vertex();
    }
}

Polygon::Polygon(const vector<Vertex> &vertices)
{
   int num;
   num=vertices.size();
   _polygon.resize(num);
   for (int i=0; i<num; i++)
       _polygon[i]=vertices[i];
}

Polygon::Polygon(queue<Vertex> &vertices)
{
    int num;
    num=vertices.size();
    _polygon.resize(num);
    for (int i=0; i<num; i++)
       {
        _polygon[i]=vertices.front();
        vertices.pop();
    }
}

Polygon::~Polygon()
{

}


int Polygon::NumberVertices() const
{
    return _polygon.size();
}

Vertex Polygon::GetVertex(const int &vertexPosition)const
{
    return _polygon[vertexPosition];
}

const Segment Polygon::GetEdge( const int &partenza)const
{
    Point a;
    Point da;
    da= _polygon[partenza].GetPoint();
    if (partenza==NumberVertices()-1)
        a=_polygon[0].GetPoint();
    else
        a=_polygon[partenza+1].GetPoint();
    Segment segment;
    segment=Segment(da,a);
    return segment;
}

bool Polygon::TypePolygon()
{
    Intersector intersection;
    intersection = Intersector();
    bool a=0; //convesso
    int N = NumberVertices();
    for(int i=1; i< N && a==0 ; i++ )
        {
        for(int j=(i+2)%N ; j!= (i-1)%N && a==0; j=(j+1)%N)
            {
            intersection.SetCutter(Segment(_polygon[i-1].GetPoint(),_polygon[i].GetPoint()));
            intersection.SetEdge(Segment(_polygon[((j-1)%N+N)%N].GetPoint(),_polygon[j%N].GetPoint())); //se c'è intersectionOnLine tra due lati non consecutivi è concavo
            bool ris= intersection.ComputeIntersection();
           if (ris==true && (intersection.TypeIntersection()==Intersector::IntersectionOnLine)) {

                 a=1;//concavo

            }
           }

        }
    if (a==1)
        return true; //concavo
    else
        return false; //convesso

}

Polygon::PositionSide Polygon::SidePoint(const Segment &segment, const Point &point)
{
    PositionSide position;
    position=Left;
    Point p2;
    Point p0;
    Point p1;
    p0=segment.GetFrom();
    p1=segment.GetTo();
    p2=point;
    Point v;
    Point w;
    Point k;
    v=Point(p1.GetX()-p0.GetX(), p1.GetY()-p0.GetY());
    w=Point(p2.GetX()-p0.GetX(), p2.GetY()-p0.GetY());
    k=Point (p2.GetX()-p1.GetX(), p2.GetY()-p1.GetY());
    Matrix2d controllo;
    controllo.col(0)=v.getCoordinate();
    controllo.col(1)=w.getCoordinate();
    double z;
    z= controllo.determinant();
    if (z>tolerance)
    {
        position=Left;
        return position;
    }
    else if (z<-tolerance)
    {
        position=Right;
        return position;
    }
    else if (v.GetX()*w.GetX()<0 || w.GetY()*v.GetY()<0)
    {
        position=Behind;
        return position;
    }
    else if (v.ComputeNorm2()<w.ComputeNorm2())
    {
        position =Forward;
        return position;
    }
    else if (w.ComputeNorm2()<=tolerance && w.ComputeNorm2()>= - tolerance)
    {
        position = Begin;
        return position;
    }
    else if (k.ComputeNorm2()<=tolerance && k.ComputeNorm2()>= - tolerance)
    {
        position=End;
        return position;
    }
    else
    {
        position =In;
        return position;
    }
}

bool Polygon::PuntoInterno(const Point &punto)
{
    bool interno=true;
    PositionSide posizione;
    int n=_polygon.size();
    if (TypePolygon()==false)
        for (int i=0; i<n; i++)
        {
            posizione=SidePoint(GetEdge(i), punto);
            if (posizione==PositionSide::Right || posizione==PositionSide::Behind|| posizione==PositionSide::Forward || posizione==PositionSide::Begin || posizione== PositionSide::End)
                interno=false;
        }
    return interno;
}

Polygon &Polygon::operator =(const Polygon &polygon)
{
    int n=polygon._polygon.size();
    _polygon.resize(n);
    for (int i=0; i<n; i++)
    {
        _polygon[i]=polygon._polygon[i];
    }
    return *this;
}

};

File: cuttingPolygonsPython/src/polygon.py
import numpy as np
import src.point as point
import src.vertex as vertex
import src.segment as segment
import src.intersector as intersector


class PositionSide:  #funziona come enum, classe con delle variabili
    Left=0
    Right = 1
    Forward = 2
    Behind = 3
    Begin = 4
    End = 5
    In = 6

class Polygon:
    def __init__(self, vertices):

        self._polygon = np.copy(vertices) #con questo fa solo la copia ma non cambia se cambio vertices

        #Per Python gestitsco sia vettori che code con copy erchè per lui sono uguali

    def numberVertices(self):
        return len(self._polygon) #è come il size di C++

    def getVertex(self, vertexPosition):
        return self._polygon[vertexPosition]

    def getEdge(self, start):
        to = point.Point()

        if start == self.numberVertices()-1:
            to = self._polygon[0].getPoint()
        else:
            to = self._polygon[start+1].getPoint()
        s = segment.Segment()
        s.setFromPoint(self._polygon[start].getPoint(), to)
        return s

    def typePolygon(self):
        intersection = intersector.Intersector()
        a= 0 #convesso
        N= self.numberVertices()
        i = 1
        while i<N and a==0:
            j=(i+2)%N
            while j!= (i-1)%N and a==0:
                s1= segment.Segment()
                from1=self._polygon[i-1].getPoint()
                to1=self._polygon[i].getPoint()
                from2=self._polygon[((j-1)%N+N)%N].getPoint()
                to2=self._polygon[j%N].getPoint()
                s1.setFromPoint(from1,to1)
                s2 = segment.Segment()
                s2.setFromPoint(from2,to2)
                intersection.setCutter(s1)
                intersection.setEdge(s2)
                ris= intersection.computeIntersection()
                if ris==True and (intersection.getTypeIntersection() == intersector.Type.IntersectionOnLine):
                    a=1 #concavo
                j=(j+1)%N
            i=i+1
        if a==1:
            return True #concavo
        else:
            return False #convesso

    def sidePoint(self, segment, pointP):
       position= PositionSide.Left
       p0= segment.getFrom()
       p1 = segment.getTo()
       p2=pointP
       v = point.Point()
       w = point.Point()
       k = point.Point()
       v.setcoordinate(p1.getX() - p0.getX(), p1.getY() - p0.getY())
       w.setcoordinate(p2.getX() - p0.getX(), p2.getY() - p0.getY())
       k.setcoordinate(p2.getX() - p1.getX(), p2.getY() - p1.getY())
       controll = np.array([v.getcoordinates(), w.getcoordinates()])
       z = np.linalg.det(controll)
       if z > segment.TOLERANCE:
           position = PositionSide.Left
           return position
       elif z < - segment.TOLERANCE:
           position = PositionSide.Right
           return position
       elif v.getX() * w.getX() < 0 or w.getY() * v.getY() < 0:
           position = PositionSide.Behind
           return position
       elif v.norm2() < w.norm2():
           position = PositionSide.Forward
           return position
       elif w.norm2() == -segment.TOLERANCE:
           position =PositionSide.Begin
           return position
       elif k.norm2() == -segment.TOLERANCE:
           position =PositionSide.End
           return position
       else:
           position = PositionSide.In
           return position

    def insidePoint(self , point2):
        inside= True

        n = len(self._polygon)
        if self.typePolygon() == False:
            i=0
            while i<n:
                position2 = self.sidePoint(self.getEdge(i),point2)
                if position2 == segment.PositionSide.Right or position2== segment.PositionSide.Behind or position2 ==segment.PositionSide.Forward or position2==  segment.PositionSide.Begin or position2== segment.PositionSide.End:
                    inside=False
                    i=i+1
                else:
                    i=i+1
            return inside








